tests/README.md

```diff
@@ -0,0 +1,1510 @@
+- Producing a list with exactly one student
+- Producing a list with ten students
+- Adding a student to the empty list
+- Adding a student to the list which contains only one other student
+- Adding a student to the list which contains not ten other students
+- Adding a student to the list which contains already ten students
+
+In general test should focus on boundary conditions with 0, 1 and n.
+
+<a name="time"></a>
+#### Time
+There are several things which should be considered regarding the wall clock time. If some portion of code
+rely on time for example a timestamp. The unit test may work 1 or 2 times but will break in the future because time itself is
+not under the control of the test and developer. Therefore, the standard time sources must be faked with more controllable ones
+to make the tests repeatable (FI**R**ST).
+
+An other aspect are timezones which are normally not a problem in PHP, but there are some edge cases with switching hours which should
+be considered. A short example from [stackoverflow](https://stackoverflow.com/a/19004000) illustrates the error in a simple way.
+
+| Europe/Warsaw time  |   offset |  UTC                |    php2utc conversion |  php offset  |
+| ------------------- | -------- | ------------------- | --------------------- | ------------ |
+| 2013-10-27 01:00:00 |    +2    | 2013-10-26 23:00:00 |   2013-10-26 23:00:00 | +2           |
+| 2013-10-27 01:30:00 |    +2    | 2013-10-26 23:30:00 |   2013-10-26 23:30:00 | +2           |
+| 2013-10-27 02:00:00 |    +2    | 2013-10-27 00:00:00 |   2013-10-27 01:00:00 | **+1**       |
+| 2013-10-27 02:30:00 |    +2    | 2013-10-27 00:30:00 |   2013-10-27 01:30:00 | **+1**       |
+| 2013-10-27 02:59:00 |    +2    | 2013-10-27 00:59:00 |   2013-10-27 01:59:00 | **+1**       |
+
+3am -> 2am .....................................summer time changes to standard(winter) time @3am we subtract 1h so 3am becomes 2am
+
+| Europe/Warsaw time  |   offset |  UTC                  |    php2utc conversion  |  php offset  |
+| ------------------- | -------- | --------------------- | ---------------------- | ------------ |
+| 2013-10-27 02:00:00 |    +1    | 2013-10-27 01:00:00   | 2013-10-27 01:00:00    | +1           |
+| 2013-10-27 02:30:00 |    +1    | 2013-10-27 01:30:00   | 2013-10-27 01:30:00    | +1           |
+| 2013-10-27 03:00:00 |    +1    | 2013-10-27 02:00:00   | 2013-10-27 02:00:00    | +1           |
+| 2013-10-27 03:30:00 |    +1    | 2013-10-27 02:30:00   | 2013-10-27 02:30:00    | +1           |
+
+
+<a name="test-examples"></a>
+## Test Examples
+This test section will show good and bad unit tests in ILIAS.
+
+<a name="negative-examples"></a>
+### Negative Examples
+All the examples are shown in this section have some different problems which will be explained in detail. However
+it is really important to understand that only the code is wrong. The statements made are never about a developers skills and how they are now
+in the present. "It's just code!"
```

Comment:
Yes indeed the most problematic part is the absence of the tests. Of course I don't want to remember other developer of "suboptimal" test they wrote, but I think to write such tests is just a way to create better ones in the future.

If someone starts to dislike an example in this guide, they are free to contribute a better one because there is always room for improvements especially after some time has passed.

The reason why I put the "positive examples with room for improvement" before the "positive examples" is that the reader should finish the documents with the good examples in mind instead of the ones with "room for improvement".

Should I rename the chapter to something more positive?

Reply:
I see your point regarding that positive examples, I'm fine leaving them at their place.

Maybe we rename the `Negative Examples` to `Suggestions for Improvements by Example`. I just realized that this is the only way, how one could get value out of a section like this: If it shows common problems and their solutions to them. I mean: if this section would show stuff that only appears once in the code base, has little chance of reappearing or little chance for generalisation, than we would not need to show it here. So to add any value, the section needs to suggest fixes to common problems in (our) tests. Besides our disagreements on specifics in the examples shown here, the headlines in this chapter seem fine in this regard.